### DISCUSSION TITLE
Issues with tfswitch and latest versions of AWS provider

### DISCUSSION THREAD

**Discussion by thawkins83** [warrensbox/terraform-switcher#514] [Q&A]
I started getting this error the other week and search in discussions & Google  to see if it has been posted.  Any AWS provider newer than 5.70 (this is the one I know still works) will occasionally give me the following error.

```
Acquiring state lock. This may take a few moments...
╷
│ Error: Failed to load plugin schemas
│ 
│ Error while loading schemas for plugin components: Failed to obtain provider schema: Could not load the schema for provider registry.terraform.io/hashicorp/aws: failed to instantiate provider
│ "registry.terraform.io/hashicorp/aws" to obtain schema: Unrecognized remote plugin message: 
│ Failed to read any lines from plugin's stdout
│ This usually means
│   the plugin was not compiled for this architecture,
│   the plugin is missing dynamic-link libraries necessary to run,
│   the plugin is not executable by this process due to file permissions, or
│   the plugin failed to negotiate the initial go-plugin protocol handshake
│ 
│ Additional notes about plugin:
│   Path: .terraform/providers/registry.terraform.io/hashicorp/aws/5.74.0/darwin_arm64/terraform-provider-aws_v5.74.0_x5
│   Mode: -rwxr-xr-x
│   Owner: <redacted>
│   Group: <redacted>
│   MachO architecture: CpuArm64 (current architecture: arm64)
│ ..
╵

```

I can't find much about this.  I have tested using tfswitch from brew and terraform from hashicorp tap in brew. I'm using version 1.9.8 of terraform.

I too worked at Pearson and that's how I discovered this app.

**Comment by yermulnik** [comment]
`tfswitch` has nothing to do with Terraform providers. `tfswitch` is pertained to Terraform (and OpenTofu) binary only.

My hint would be for you to try and delete `.terraform/providers/` dir recursively and then re-init Terraform (`terraform init`) so that it attempts to fresh download required providers.

**Reply by thawkins83** [reply]
I've already deleted providers multiple times. I've deleted binaries from .terraform.version, still get same results. Again, this only occurs when using the tfswitch downloads of terraform binary. 

I guess my only option is to stop using tfswitch based on your answer.

**Reply by yermulnik** [reply]
> Again, this only occurs when using the tfswitch downloads of terraform binary.

And does not occur when using TF binary downloaded manually or installed from Hashcorp's Brew tap?
Would you mind providing output of:
* `uname -a`
* `which terraform`
* `terraform version`
* `file $(realpath $(which terraform))`
* `tfswitch --version`
* `printenv | grep -E ^TF_`
* `file .terraform/providers/registry.terraform.io/hashicorp/aws/5.74.0/darwin_arm64/terraform-provider-aws_v5.74.0_x5`

**Reply by yermulnik** [reply]
Also please provide full command line and output of `tfswitch` command that you run to install/switch TF version.

ps: it might be that `tfswitch` mistakenly installs `amd64` version of TF instead of `arm64` (if your on Apple Silicon M1 (or whatever it is called in Mac world)). if it is that, then you've found a bug and we will need to figure how to fix that (despite you're the only one who have reported such an issue so far 😕).

**Reply by yermulnik** [reply]
Interestingly that coincidentally I'm seeing misbehavior reports around Terraform, latest AWS provider (5.74.0) and Mac on ARM chips running AMD via Rosetta 🤔 So it might be something odd getting in the way.
Please post what I asked for above when you get a chance so that we can figure out whether it's something that `tfswitch` is doing incorrectly (e.g. downloading AMD version of Terraform binary instead of ARM) or there's a larger issue that we may or may not be able to remediate within `tfswitch` ecosystem.

**Reply by MatthewJohn** [reply]
> I'm seeing misbehavior reports around Terraform, latest AWS provider (5.74.0) and Mac on ARM chips running AMD via Rosetta

Do you have a link to a report at all?
Though, if this is about AMD64 binary on ARM64, we can at least see from the error that it _seems_ to be the ARM64 version of the provider 🤔 (unless the problem is AMD64 version of Terraform running using Rosetta and incorrectly downloading the ARM64 provider version (and thus the error reports)) - Maybe a bit of a long-shot though :P

**Reply by yermulnik** [reply]
> Do you have a link to a report at all?

Do you happen to be a member of HangOps Slack?
That's not exact or even similar, though the conditions are quite the same:
![image](https://github.com/user-attachments/assets/94b405f8-4bae-415e-868a-168c538628c4)
And what's Martin (former TF core dev) is saying matches more or less what my assumption is:
![image](https://github.com/user-attachments/assets/d991800c-854d-460d-a37c-66f923cd3814)

> (unless the problem is AMD64 version of Terraform running using Rosetta and incorrectly downloading the ARM64 provider version (and thus the error reports))

That's my shot too.
Let's see what the outputs are from the commands that I asked @thawkins83 for.
Maybe we'd want to have something else provided to help investigation, wdyt?

**Reply by thawkins83** [reply]
> And does not occur when using TF binary downloaded manually or installed from Hashcorp's Brew tap?
> Would you mind providing output of:

This does not occur when using Hashicorp Brew tap.

$ uname -a
Darwin MACWFLWCD6GL0 23.6.0 Darwin Kernel Version 23.6.0: Wed Jul 31 20:48:52 PDT 2024; root:xnu-10063.141.1.700.5~1/RELEASE_ARM64_T6020 arm64 arm Darwin

$ which terraform
/Users/<redacted>/bin/terraform

$terraform version
Terraform v1.9.8
on darwin_arm64
\+ provider registry.terraform.io/hashicorp/aws v5.74.0

$ tfswitch --version
Version: v1.2.3

$ file $(realpath $(which terraform))
/Users/<redacted>/.terraform.versions/terraform_1.9.8: Mach-O 64-bit executable arm64

$ printenv | grep -E ^TF_
<empty>

$ file .terraform/providers/registry.terraform.io/hashicorp/aws/5.74.0/darwin_arm64/terraform-provider-aws_v5.74.0_x5
.terraform/providers/registry.terraform.io/hashicorp/aws/5.74.0/darwin_arm64/terraform-provider-aws_v5.74.0_x5: Mach-O 64-bit executable arm64


> Also please provide full command line and output of tfswitch command that you run to install/switch TF version.

$ tfswitch
✔ 1.9.8 *recent
17:58:16.691 INFO Downloading "https://releases.hashicorp.com/terraform/1.9.8/terraform_1.9.8_darwin_arm64.zip"  
17:58:16.691 INFO Downloading to "/Users/<redacted>/.terraform.versions/terraform_1.9.8_darwin_arm64.zip"  
17:58:21.027 INFO 25745690 bytes downloaded  
17:58:21.028 INFO Downloading "https://releases.hashicorp.com/terraform/1.9.8/terraform_1.9.8_SHA256SUMS"  
17:58:21.028 INFO Downloading to "/Users/<redacted>/.terraform.versions/terraform_1.9.8_SHA256SUMS"  
17:58:21.072 INFO 1378 bytes downloaded  
17:58:21.072 INFO Downloading "https://releases.hashicorp.com/terraform/1.9.8/terraform_1.9.8_SHA256SUMS.72D7468F.sig"  
17:58:21.072 INFO Downloading to "/Users/<redacted>/.terraform.versions/terraform_1.9.8_SHA256SUMS.72D7468F.sig"  
17:58:21.108 INFO 566 bytes downloaded  
17:58:21.112 INFO Verifying signature of checksum file...  
17:58:21.114 INFO Checksum file signature verification successful.  
17:58:21.127 INFO Deleting "/Users/<redacted>/.terraform.versions/terraform_1.9.8_SHA256SUMS"  
17:58:21.140 INFO Deleting "/Users/<redacted>/.terraform.versions/terraform_1.9.8_SHA256SUMS.72D7468F.sig"  
17:58:21.512 INFO Switched Terraform to version "1.9.8"  

> Interestingly that coincidentally I'm seeing misbehavior reports around Terraform, latest AWS provider (5.74.0) and Mac on ARM chips running AMD via Rosetta 🤔 So it might be something odd getting in the way.

Best I can remember, this has happened since AWS provider 5.71.0 was release, which they removed. I don't think I used 5.72.0. And I know this happened with 5.72.1, 5.73.0, & 5.74.0

> Do you happen to be a member of HangOps Slack?

I am not

> That's not exact or even similar, though the conditions are quite the same:

Yeah, that seems totally different.


If I missed anything, just lmk. I didn't ask on AWS provider forum because I didn't see anyone else having the same issue and it was only through the binary in .terraform.versions

**Reply by yermulnik** [reply]
@thawkins83 Thank you.

> This does not occur when using Hashicorp Brew tap.

Could you please also provide `terraform version` and `file /path/to/terraform/binary` installed via Hashicorp Brew tap?
Probably something like below should do the thing:
* `brew install terraform` # if not yet
* `$(brew --prefix)/bin/terraform version`
* `file $(realpath $(brew --prefix)/bin/terraform)`

UPD: from what I can see Hashicorp TF tap installes `1.9.8` TF binary from the same location as `tfswitch` does — https://github.com/hashicorp/homebrew-tap/blob/master/Formula/terraform.rb#L15C10-L15C88 — so I'm expecting binaries must not differ:
```shell
diff $(realpath $(which terraform)) $(realpath $(brew --prefix)/bin/terraform)
```

**Reply by thawkins83** [reply]
> Could you please also provide `terraform version` and `file /path/to/terraform/binary` installed via Hashicorp Brew tap? Probably something like below should do the thing:
> 
> * `brew install terraform` # if not yet
> * `$(brew --prefix)/bin/terraform version`
> * `file $(realpath $(brew --prefix)/bin/terraform)`

```  shell
$ brew install hashicorp/tap/terraform # brew install terraform installs 1.5.8
==> Auto-updating Homebrew...
Adjust how often this is run with HOMEBREW_AUTO_UPDATE_SECS or disable with
HOMEBREW_NO_AUTO_UPDATE. Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
==> Auto-updated Homebrew!
Updated 2 taps (hashicorp/tap and homebrew/core).
==> New Formulae
dipc

You have 2 outdated formulae installed.

==> Fetching hashicorp/tap/terraform
==> Downloading https://releases.hashicorp.com/terraform/1.9.8/terraform_1.9.8_darwin_arm64.zip
Already downloaded: /Users/<redacted>/Library/Caches/Homebrew/downloads/7bff84d284f6ca2902835add96972f200d6650bd44971369dd1cc75b51d296ab--terraform_1.9.8_darwin_arm64.zip
==> Installing terraform from hashicorp/tap
🍺  /opt/homebrew/Cellar/terraform/1.9.8: 5 files, 83.9MB, built in 1 second
==> Running `brew cleanup terraform`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).

$ $(brew --prefix)/bin/terraform version
Terraform v1.9.8
on darwin_arm64

$ file $(realpath $(brew --prefix)/bin/terraform)
/opt/homebrew/Cellar/terraform/1.9.8/bin/terraform: Mach-O 64-bit executable arm64
```

> ```shell
> diff $(realpath $(which terraform)) $(realpath $(brew --prefix)/bin/terraform)
> ```

```shell
$ diff $(realpath $(which terraform)) $(realpath $(brew --prefix)/bin/terraform)
# nothing returned
```

One thing I did notice, is that I need to delete the version in .terraform.versions, otherwise `realpath $(which terraform)` returns that download.

**Reply by yermulnik** [reply]
> One thing I did notice, is that I need to delete the version in .terraform.versions, otherwise `realpath $(which terraform)` returns that download.

That's the idea: compare `terraform` binaries installed by `tfswitch` and by Homebrew.

> ```
> $ diff $(realpath $(which terraform)) $(realpath $(brew --prefix)/bin/terraform)
> # nothing returned
> ```

Did you run this command **before** you deleted the one from `.terraform.versions` dir or **after**?
In the latter case please re-run `tfswitch` to have it installed TF 1.9.8 and run `diff $(realpath $(which terraform)) $(realpath $(brew --prefix)/bin/terraform)` again (and additionally `file $(realpath $(which terraform)) $(realpath $(brew --prefix)/bin/terraform)` just in case).
In the former case, granted both `tfswitch` and Homebrew install that same `Mach-O 64-bit executable arm64` binary, I'm out of ideas. Maybe also try and run `terraform` command that fails loading plugin schemas using Terraform binary installed by each of `tfswitch` and Homebrew:
* `$(realpath $(which terraform)) ...`
* `$(realpath $(brew --prefix)/bin/terraform) ...`

**Reply by yermulnik** [reply]
I guess if these two TF binaries are the same, you should be getting failures running either of them. Else it's something specific to your local env or OS (e.g. some security oddness applied by macOS or similar). Ex https://github.com/warrensbox/terraform-switcher/issues/307 (and linked issue from TF repo) or https://github.com/warrensbox/terraform-switcher/issues/310

**Reply by thawkins83** [reply]
> Did you run this command before you deleted the one from .terraform.versions dir or after?
In the latter case please re-run tfswitch to have it installed TF 1.9.8 and run diff $(realpath $(which terraform)) $(realpath $(brew --prefix)/bin/terraform) again (and additionally file $(realpath $(which terraform)) $(realpath $(brew --prefix)/bin/terraform) just in case).
In the former case, granted both tfswitch and Homebrew install that same Mach-O 64-bit executable arm64 binary, I'm out of ideas. Maybe also try and run terraform command that fails loading plugin schemas using Terraform binary installed by each of tfswitch and Homebrew:

Yes, it was run before deleting and returned nothing, meaning they're identical. I'll try running with Hashicorp more and see if I can get the issue or not.

**Reply by yermulnik** [reply]
Once you confirm Homebrew's binary has no said issue, you may want to copy it over the one installed by `tfswitch` and validate it to see whether this change affects it.

**Reply by thawkins83** [reply]
I was finally able to get the error to occur with Hashcorp install.  Looks like you guys are off the hook and it may be AWS provider issue after all.

**Reply by yermulnik** [reply]
Thanks for the feedback 👍🏻
I'd say it may be TF **or** TF AWS provider.
Please post here the Hashicorp issue link for reference if you get to create one. Thanks.

**Reply by yermulnik** [reply]
I'm closing this discussion as we can't help here any further 😞
We'll re-open if there are more posts coming in. Any objections?

**Reply by thawkins83** [reply]
This post has the information.

https://github.com/hashicorp/terraform-provider-aws/issues/39523#issuecomment-2423107701

It appears to be an ARM based issue on Macs and Rosetta.

**Reply by yermulnik** [reply]
Thank you. I saw that issue while was looking for issues similar to what you've hit. Though it was, similarly to what I linked earlier, of another kind while pertaining to the same conditions so I didn't link it to this thread.
Hope they will rectify this issue asap 🤞🏻
